### PR TITLE
Open the scripts in zfs_snapshot

### PR DESCRIPTION
**Pull request by ChaoyuZhang**
As the scripts(`zfs_snapshot_008_neg`, `zfs_snapshot_009_pos`) in zfs_snapshot can run successfully on my linux,  open these two scripts to see if they pass the buildbot.

### CODE REVIEW

**Comment by behlendorf** [comment]
@ChaoyuZhang these tests were disabled because they can trigger a deadlock by creating ZFS pools on files within other ZFS datasets.  This isn't entirely safe.  Before these tests can be enabled they need to be reworked slightly to avoid this kind of nesting.

**Comment by ChaoyuZhang** [comment]
@behlendorf  The ninth script copy data from zpool_upgrade which are used to construct a pool of a given
version. but this script will clear the filesystem and snapshot. Is it will affect the scripts in zpool_upgrade?
Can we copy other data to the filesystem? By the way, the command `DIRCMP` is not found, can we replace this command with `diff` ?

**Comment by behlendorf** [comment]
This shouldn't have any effect in the zpool_upgrade tests.  The tarball is just being use as a source for some data for the purposes of the test.  

We can use `diff -r` for dircmp but the scripts may need to be updated slightly for Linux.  It would be best to write this in a way which works on both platforms but it would be OK to put this in an is_linux conditional.

File: tests/zfs-tests/tests/functional/cli_root/zfs_snapshot/zfs_snapshot_009_pos.ksh [hunk 1]
```diff
@@ -84,9 +84,15 @@ for i in 1 2 3; do
 done
 log_note "verify snapshot contents"
 for ds in $datasets; do
-	status=$($DIRCMP /$ds /$ds/.zfs/snapshot/snap | $GREP "different")
-	[[ -z $status ]] || log_fail "snapshot contents are different from" \
-	    "the filesystem"
+	if is_linux; then 
+		status=$($DIFF -r /$ds /$ds/.zfs/snapshot/snap | $GREP "different")
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/cli_root/zfs_snapshot/zfs_snapshot_009_pos.ksh]: This won't work as is. `diff -r` doesn't output the word different when two directory trees different.  However it does return 1 when directories are different and 0 when they are the same.  For example:

```
$ mkdir foo
$ mkdir bar
$ touch foo/a
$ touch foo/b
$ touch bar/a
$ diff -r foo/ bar/; echo $?
Only in foo/: b
1
$ touch bar/b
$ diff -r foo/ bar/; echo $?
0
```

**Comment by ChaoyuZhang** [comment]
According to the script, create fs and copy data, then create snapshot and compare the content. However, there are nothing in the snap. We must create a clone and all the data  can be seen in the clone. Test results show as below:
```
[root@A22832414_00 zfs_set]# zfs create test/test5
[root@A22832414_00 zfs_set]# cp /home/zcy/TestData/* /test/test5
[root@A22832414_00 zfs_set]# zfs snapshot test/test5@snap1
[root@A22832414_00 zfs_set]# cd /test/test5/.zfs/snapshot/snap1/
[root@A22832414_00 snap1]# ls
[root@A22832414_00 snap1]# zfs clone test/test5@snap1 test/clone
[root@A22832414_00 snap1]# cd /test/clone
[root@A22832414_00 clone]# ls
0.5m.txt  1k.txt  1m.txt
[root@A22832414_00 clone]# cd /test/test5
[root@A22832414_00 test5]# ls
0.5m.txt  1k.txt  1m.txt
```
Why do I set the `snapdir=hidden`? As the clone has a dir `.zfs`.
```
[root@A22832414_00 test5]# diff -r /home/zcy/TestData/ /test/clone; echo $?
Only in /test/clone: .zfs
1
[root@A22832414_00 test5]# zfs set snapdir=hidden test/clone
[root@A22832414_00 test5]# diff -r /home/zcy/TestData/ /test/clone; echo $?
0
```

**Comment by behlendorf** [comment]
@ChaoyuZhang I'm not observing failures with either of these test cases on CentOS 7.  I've opened #5260 which enables both of these test to get a test run through the buildbots.

**Comment by ChaoyuZhang** [comment]
@behlendorf  Thanks!  The ninth test case ran failed maybe as my linux is RedHat, I will run the test cases in CentOS in future.

**Comment by behlendorf** [comment]
Closing is favor of #5260.

**Pull request closed without merging**